actioncable/test/javascript/src/unit/subscription_test.coffee

```diff
@@ -0,0 +1,66 @@
+{module, test} = QUnit
+{consumerTest, defer} = ActionCable.TestHelpers
+
+current = {}
+
+afterEach = ->
+  current = {}
+
+module "ActionCable.Subscription", {afterEach}, ->
+  consumerTest "callbacks", ({server, consumer, assert, done}) ->
+    current.server = server
+    current.subscription = consumer.subscriptions.create "chat",
+      testData: {}
+
+      initialized: ->
+        @testData.initialized = true
+
+      connected: ->
+        @testData.connected = true
+
+      received: (data) ->
+        @testData.received ?= []
+        @testData.received.push(data)
+
+      disconnected: ->
+        @testData.disconnected = true
+
+    # trigger connected callback
+    broadcast message_type: "confirmation", ->
+      # trigger received callback
+      broadcast message: { line: "hi" }, ->
+        # trigger disconnected callback
+        server.close()
+        defer ->
+          {testData} = current.subscription
+          assert.ok testData.initialized
+          assert.ok testData.connected
+          assert.ok testData.disconnected
+          assert.deepEqual testData.received, [line: "hi"]
+          done()
+
+  consumerTest "perform", ({consumer, server, assert, done}) ->
+    current.server = server
+    current.subscription = consumer.subscriptions.create "chat",
+      connected: ->
+        @perform(publish: "hi")
+
+    server.on "message", (message) ->
+      data = JSON.parse(message)
+      assert.equal data.identifier, current.subscription.identifier
+      assert.equal data.command, "message"
+      assert.deepEqual data.data, JSON.stringify(action: { publish: "hi" })
+      done()
+
+    broadcast(message_type: "confirmation")
+
+
+broadcast = (data = {}, callback) ->
```

Comment:
Might be a bit early, but should we extract the `broadcast` function out to where the rest of the "helper" functions are?